### ISSUE TITLE
Manual implementation of var does not give the same value as torch.var

### ISSUE THREAD

**Issue by mreyarea-fy** [pytorch/pytorch#88416]
### 🐛 Describe the bug

If I try to calculate variance manually, it is never the same as the `torch.var` function. They are close, but I was expecting exactly the same value as when I compute manually the mean compared with `torch.mean`. 

This could be a problem when we try to compute our own version of batchnorm for example. I need to compute mean/var manually because I have masked values. 

I don't know if this is a bug, but why is this happening?

```
import torch

torch.set_printoptions(precision=10)
for i in range(10):
    x = torch.randn((10, 3, 100, 100))
    nelem = 10 * 100 * 100
    mean = x.sum(dim=(0, 2, 3)) / nelem
    var = torch.sum((x - mean[None, :, None, None]) ** 2, dim=(0, 2, 3)) / nelem

    torch_var = x.var(dim=(0, 2, 3), unbiased=False)
    torch_mean = x.mean(dim=(0, 2, 3))
    print(i)
    print(f"means are exactly equal {torch.all(mean == torch_mean)}")
    print(f"vars are exactly equal {torch.all(var == torch_var)}")
```
In this run for example, only the last iteration of the for loop gives exactly equal variances
```
0
means are exactly equal True
vars are exactly equal False
1
means are exactly equal True
vars are exactly equal False
2
means are exactly equal True
vars are exactly equal False
3
means are exactly equal True
vars are exactly equal False
4
means are exactly equal True
vars are exactly equal False
5
means are exactly equal True
vars are exactly equal False
6
means are exactly equal True
vars are exactly equal False
7
means are exactly equal True
vars are exactly equal False
8
means are exactly equal True
vars are exactly equal False
9
means are exactly equal True
vars are exactly equal True
```

### Versions

```
PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA GeForce GTX 1080 Ti
GPU 1: NVIDIA GeForce GTX TITAN X

Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.6.5
[pip3] torch==1.12.1
[pip3] torchmetrics==0.10.0
[pip3] torchvision==0.13.1
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] pytorch-lightning         1.6.5                    pypi_0    pypi
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torchmetrics              0.10.0                   pypi_0    pypi
[conda] torchvision               0.13.1                   pypi_0    pypi

```

**Comment by zou3519** [comment]
Use torch.allclose to compare two Tensors, not ==. Changing the above script to compare torch.allclose makes it all return True for me.

**Comment by mreyarea-fy** [comment]
Right I know that. But I was expecting the exact same numbers. In numpy I get the exact same numbers. 

To illustrate this. For example, if I write my own implementation of batch norm like the following:

```
import torch
from torch import nn

class CustomBatchNorm(nn.BatchNorm2d):

    def forward(self, input):
        self._check_input_dim(input)

        if self.training and self.track_running_stats:
            if self.num_batches_tracked is not None:
                self.num_batches_tracked += 1
                if self.momentum is None:  # use cumulative moving average
                    exponential_average_factor = 1.0 / float(self.num_batches_tracked)
                else:  # use exponential moving average
                    exponential_average_factor = self.momentum

        B, C, H, W = input.shape
        if self.training:
            n = input.shape[0] * input.shape[2] * input.shape[3]
            mean = torch.sum(input, dim=(0, 2, 3)) / n
            var = (input - mean.view(1, C, 1, 1)) ** 2
            var = var.sum(dim=(0, 2, 3)) / n

            with torch.no_grad():
                self.running_mean = exponential_average_factor * mean + (1 - exponential_average_factor) * self.running_mean
                self.running_var = exponential_average_factor * var * n / (n - 1) + \
                                   (1 - exponential_average_factor) * self.running_var

        else:
            mean = self.running_mean
            var = self.running_var

        # Norm the input
        normed = (input - mean.view(1, C, 1, 1)) / (torch.sqrt(var + self.eps)).view(1, C, 1, 1)

        # Apply affine parameters
        if self.affine:
            normed = normed * self.weight.view(1, C, 1, 1) + self.bias.view(1, C, 1, 1)
        return normed
```
And then I compare it with the `torch.nn.BatchNorm2d` like this:
```
my_bn = CustomBatchNorm(3, affine=True)
bn = nn.BatchNorm2d(3, affine=True)

# Run train
for _ in range(10):
    scale = torch.randint(1, 10, (1,)).float()
    bias = torch.randint(-10, 10, (1,)).float()
    x = torch.randn(10, 3, 100, 100) * scale + bias
    out1 = my_bn(x)
    out2 = bn(x)
    print(torch.allclose(out1, out2))
```
my output is all `False` for a run. There are some runs I get `True` values but never all of them
```
False
False
False
False
False
False
False
False
False
False
```
Both layers are initialized exactly with the same parameters, therefore the only thing that can change is the variance between runs no? As I've shown before torch.mean and the manual mean are exactly the same. However, the manual variance and the torch.var are not exactly the same. This can be propagated to why the batchnorms dont output the same?
